packages/block-editor/src/components/global-styles/typography-panel.js

```diff
@@ -252,6 +255,39 @@ export default function TypographyPanel( {
 		setFontAppearance( {} );
 	};
 
+	// Check if previous font style and weight values are available in the new font family
+	useEffect( () => {
+		const { fontStyles, fontWeights, isSystemFont } =
+			getFontStylesAndWeights( fontFamilyFaces );
+		const hasFontStyle = fontStyles?.some(
+			( { value: fs } ) => fs === fontStyle
+		);
+		const hasFontWeight = fontWeights?.some(
+			( { value: fw } ) => fw === fontWeight
+		);
+
+		// Try to set nearest available font weight
+		if ( ! hasFontWeight && fontWeight ) {
+			setFontAppearance( {
+				fontStyle,
+				fontWeight: findNearestFontWeight( fontWeights, fontWeight ),
+			} );
+		}
+
+		// Set the same weight and style values if the font family is a system font or if both are the same
+		if ( isSystemFont || ( hasFontStyle && hasFontWeight ) ) {
+			setFontAppearance( {
+				fontStyle,
+				fontWeight,
+			} );
+		}
+
+		// Reset font appearance if the font family does not have the selected font style
+		if ( ! hasFontStyle ) {
+			resetFontAppearance();
+		}
```

Comment:
@mikachan I noticed here that if `isSystemFont` is `true` and if `hasFontWeight` has a value, then if first calls `setFontAppearance` to set the font weight, but will straight away call `resetFontAppearance` and unsets the value that was just set.

The background is that I've been working on some e2e tests (#61932) that programatically set `fontWeight` a bit like this:
```js
var block = wp.blocks.createBlock('core/paragraph', {
  content: 'Hello',
  style: { typography: { fontWeight: '300' } } }
);
wp.data.dispatch('core/block-editor').insertBlock( block )
```

But that recently stopped working, and I think it's due to the `resetFontAppearance` here. To reproduce, if you run that in the console the block that's inserted ends up with no appearance styles.

I'm not familiar with the code, but thought I'd mention it.

Reply:
Thanks for highlighting, @talldan! I'll take a look and open a follow-up PR. I'm guessing the `hasFontStyle` and `hasFontWeight` combined logic needs to be adjusted.

Reply:
@talldan I've opened a follow-up here that hopefully fixes this: https://github.com/WordPress/gutenberg/pull/63215 🤞